Update Collections (#179)
hpi-swa-lab/SqueakByExample-english: Collections/Collections.tex

For the purposes of this chapter, we use the term ``Collection Hierarchy'' to mean \ct{Collection} and its 65 subclasses that are \emph{also} in the system categories labeled \scat{Collections-*}.
In this chapter, we focus mainly on the subset of collection classes shown in \figref{CollClassesList}\footnote{The group of classes that appears in the ``\ind{Blue Book}''~\cite{Gold83a} contains 17 subclasses of \ct{Collection}, and had already been redesigned several times before the \st-80 system was released.
This group of classes is often considered to be a paradigmatic example of object-oriented design.}.

In \st, when one speaks of a collection without being more specific about the kind of collection, one means an object that supports well-defined protocols for testing membership and enumerating the elements.
\emph{All} collections understand the \protind{testing} messages \mbox{\mthind{Collection}{includes:},} \mthind{Collection}{isEmpty}, and \mbox{\mthind{Collection}{occurrencesOf:}.}

> \emph{All} collections understand the \prot{enumeration} messages \mthind{Collection}{do:}, \mthind{Collection}{select:}, \mthind{Collection}{reject:} (which is the opposite of \ct{select:}) (which is like \ct{filter} in \ind{Lisp}, JavaScript, and others), \mthind{Collection}{collect:} (which is like \ct{map} in Lisp \etc), \mthind{Collection}{detect:ifNone:}, \mthind{Collection}{inject:into:} (which performs a left fold), and many more.

Reviewer: is the "(which is like filter" at the correct location, looks to me as if it refers to reject: which is not the semantic of filter:

Author: ```suggestion
\emph{All} collections understand the \prot{enumeration} messages \mthind{Collection}{do:}, \mthind{Collection}{select:} (which is like \ct{filter} in \ind{Lisp}, JavaScript, and others), \mthind{Collection}{reject:} (which is the opposite of \ct{select:}), \mthind{Collection}{collect:} (which is like \ct{map} in Lisp \etc), \mthind{Collection}{detect:ifNone:}, \mthind{Collection}{inject:into:} (which performs a left fold), and many more.
```

\begin{code}{@TEST | anArray | }
anArray := Array new: 5.
anArray at: 1 put: 4. --> 4
anArray at: 2 put: 3 / 2.
anArray at: 3 put: 'ssss'.
anArray at: 1 --> 4
\end{code}
\noindent

> Take care that \ct{at:put:} returns the argument and not the receiver, so you often need to build a cascade \mthind{Object}{yourself}!

Reviewer: "Take care" -> "Beware"

Author: ```suggestion
Beware that \ct{at:put:} returns the argument and not the receiver, so you often need to build a cascade \mthind{Object}{yourself}!
```

\cmindex{Array}{copy}

\noindent
Be careful with code that modifies array literals!
The compiler allocates space just once for them.
Unless you copy the array, it will be a \emphind{read-only object} and raise a \clsind{ModificationForbidden} error every time you try to alter it.

> Dynamically created arrays do not have this problem.

Reviewer: "problem" -> "constraint" (its a feature after all)

Author: ```suggestion
Dynamically created arrays do not have this constraint.
```

(1 to: 5) collect: [:each | each * 2] --> #(2 4 6 8 10)
\end{code}

\paragraph{Shortcuts.}

> If you look up the implementation of \ct{collect:}, \ct{do:}, \autc in any collection class, you will find that they evaluate the argument block by sending the message \mthind{BlockClosure}{value:} to it.

Reviewer: \mthind{BlockClosure}{value:} is the method, so it should just be "value:"

Author: What do you mean? In the PDF, this will just appear as `value:`. `\mthind` is for the correct index only.

(1 to: 5) collect: [:each | each * 2] --> #(2 4 6 8 10)
\end{code}

\paragraph{Shortcuts.}

If you look up the implementation of \ct{collect:}, \ct{do:}, \autc in any collection class, you will find that they evaluate the argument block by sending the message \mthind{BlockClosure}{value:} to it.
That means that you actually can pass any other object that understands the message \ct{value:} to \ct{collect:}, \ct{do:}, \etc*.
In particular, this includes the classes \clsind{Symbol} and \clsind{MessageSend}.
Thus, you could also write the above expression as follows:
\begin{code}{@TEST}
#(2 -3 4 -35 4 -11) collect: #abs --> #(2 3 4 35 4 11)

> #(2 -3 4 -35 4 -11) collect: (Message selector: #abs) --> #(2 3 4 35 4 11)

Author: ```suggestion
```
